packages/teleport/src/components/FilterByLabelList/FilterByLabelList.tsx

```diff
@@ -0,0 +1,132 @@
+/**
+ * Copyright 2021 Gravitational, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, { useState, useMemo } from 'react';
+import { useLocation } from 'react-router';
+import { Option } from 'shared/components/Select';
+import SelectFilter from 'teleport/components/SelectFilter';
+import history from 'teleport/services/history';
+
+export default function FilterableList({
+  data = [],
+  TableComponent,
+  ...tableProps
+}: Props) {
+  const { search, pathname } = useLocation();
+  const labelsFromQuery = useMemo<Option[]>(() => {
+    const searchParams = new URLSearchParams(search);
+    const query = searchParams.get('labels');
+    if (!query || query === 'null') {
+      return [];
+    }
+
+    return decodeURIComponent(query)
+      .split(',')
+      .map(label => ({ value: label, label }));
+  }, []);
+
+  const labels = useMemo<Option[]>(() => makeLabelOptions(data), []);
+  const [selectedLabels, setSelectedLabels] = useState<Option[]>(
+    labelsFromQuery
+  );
+  const [filteredData, setFilteredData] = useState(() =>
+    filterData(data, labelsFromQuery)
+  );
+
+  function onFilterApply(labels: Option[]) {
+    const filtered = filterData(data, labels);
+
+    setFilteredData(filtered);
+    setSelectedLabels(labels);
+    updateUrlQuery(labels, pathname);
+  }
+
+  function onLabelClick(label: string) {
+    let copy = selectedLabels;
+    const index = selectedLabels.findIndex(tag => tag.value === label);
+
+    if (index > -1) {
+      // remove the label
+      copy.splice(index, 1);
```

Comment:
It also mutates original `selectedLabels` here, I would choose to filter out that index

Reply:
i needed a new reference to let react know something changed